core/preconditioner/batch_jacobi.cpp

```diff
@@ -0,0 +1,177 @@
+
+template <typename ValueType, typename IndexType>
+void Jacobi<ValueType, IndexType>::generate_precond(
+    const BatchLinOp* const system_matrix)
+{
+    using unbatch_type = gko::matrix::Csr<ValueType, IndexType>;
+    // generate entire batch of factorizations
+    auto exec = this->get_executor();
+
+    if (parameters_.max_block_size == 1u) {
+        // External generate does nothing in case of scalar block jacobi (as the
+        // whole generation is done inside the solver kernel)
+        num_blocks_ = system_matrix->get_common_size()[0];
+        blocks_ = gko::array<ValueType>(exec);
+        parameters_.block_pointers = gko::array<IndexType>(exec);
+        return;
+    }
+
+    const matrix_type* sys_csr =
+        dynamic_cast<const matrix_type*>(system_matrix);
+    std::shared_ptr<const matrix_type> sys_csr_shared_ptr{};
+
+    if (!sys_csr) {
+        sys_csr_shared_ptr = gko::share(matrix_type::create(exec));
+        as<ConvertibleTo<const matrix_type>>(system_matrix)
+            ->convert_to(sys_csr_shared_ptr.get());
+        sys_csr = sys_csr_shared_ptr.get();
+    }
+
+    const auto num_batch = sys_csr->get_num_batch_items();
+    const auto num_rows = sys_csr->get_common_size()[0];
+    const auto num_nz = sys_csr->get_num_elements_per_item();
+
+    // extract the first matrix, as a view, into a regular Csr matrix.
+    const auto unbatch_size =
+        gko::dim<2>{num_rows, sys_csr->get_common_size()[1]};
+    auto sys_rows_view = array<IndexType>::const_view(
+        exec, num_rows + 1, sys_csr->get_const_row_ptrs());
+    auto sys_cols_view = array<IndexType>::const_view(
+        exec, num_nz, sys_csr->get_const_col_idxs());
+    auto sys_vals_view =
+        array<ValueType>::const_view(exec, num_nz, sys_csr->get_const_values());
+    auto first_sys_csr = gko::share(unbatch_type::create_const(
+        exec, unbatch_size, std::move(sys_vals_view), std::move(sys_cols_view),
+        std::move(sys_rows_view)));
+
+    if (parameters_.block_pointers.get_data() == nullptr) {
+        this->detect_blocks(num_batch, first_sys_csr.get());
+        exec->synchronize();
+        blocks_cumulative_storage_.resize_and_reset(num_blocks_ + 1);
+    }
+
+    // cumulative block storage
+    exec->run(jacobi::make_compute_cumulative_block_storage(
+        num_blocks_, parameters_.block_pointers.get_const_data(),
+        blocks_cumulative_storage_.get_data()));
+
+    blocks_.resize_and_reset(this->compute_storage_space(num_batch));
+
+    exec->run(jacobi::make_find_row_block_map(
+        num_blocks_, parameters_.block_pointers.get_const_data(),
+        row_block_map_info_.get_data()));
+
+    // Note: Row-major order offers advantage in terms of
+    // performance in both preconditioner generation and application for both
+    // reference and cuda backend.
+    // Note: The pattern blocks in block_pattern are
+    // also stored in a similar way.
+
+    // array for storing the common pattern of the diagonal blocks
```

Comment:
'pattern' is a bit vague, judging from the implementation, this stores the nnz indices of the common sparsity pattern for each diagonal block. So maybe `block_nnz_idxs` might be better suited?